### PR TITLE
Composable docker images with Nix

### PR DESCRIPTION
**Pull request by lucabrunox**
So I wanted a Nix function to create docker images. That's easy, just make a tarball of the closure of a derivation, and it's ready to be imported.

The problem is: how do you add more content to an existing image, by using layers?

So here's an attempt to provide a solution, even if it's very ugly.

Let's start with a simple bash container:

```
$ nix-build -E 'with import <nixpkgs> {}; dockerImage { drv = bashInteractive; drvCommand = "${bashInteractive}/bin/bash"; }'
# cat result/image.tar.gz|docker import - nixbash
# docker run -t -i nixbash /bin/bash
bash-4.3# 
# docker history nixbash
IMAGE               CREATED             CREATED BY          SIZE
4055d6dc456e        30 minutes ago                          186.4 MB
```

Now let's create another image with python only:

```
$ nix-build -E 'with import <nixpkgs> {}; dockerImage { drv = python; drvCommand = "${python}/bin/python"; }'
# cat result/image.tar.gz|docker import - nixpython
# docker run -t -i nixpython /bin/python
Python 2.7.10 (default, Jan 01 1970, 00:00:01) 
...
# docker history nixpython
IMAGE               CREATED             CREATED BY          SIZE
4ed68027aae5        10 seconds ago                          223.5 MB
```

**Now it comes the magic**. Let's layer that python image on top of bash:

```
$ nix-build -E 'with import <nixpkgs> {}; dockerImage { drv = python; drvCommand = "${python}/bin/python"; baseImage = "nixbash"; }'
# cd result
# ./docker-build.sh -t nixpython2 --no-cache .
# docker history nixpython2
IMAGE               CREATED             CREATED BY                                      SIZE
51b3911260fc        17 seconds ago      /bin/sh -c hostip=$(/nix/store/wa3fimm2m26fwh   46.57 MB
4055d6dc456e        24 minutes ago                                                      186.4 MB
```

**Note**: the baseImage is there only for convenience. One might as well generate a `Dockerfile` on the fly, and use the same `image.tar.gz`. I haven't done that, but it's easy to do.

In other words, we're able to take arbitrary nix containers and layer multiple of them with minimal overhead.

Now that 184mb is because for some reason gcc is pulled in with a lot of other stuff, most of which are useless for doing the layer merge. The python closure is only 36mb, so... there's a lot of opportunity to optimize.
Just think a debian docker is around 150mb, so we aren't that huge either.

cc @offlinehacker @datakurre @domenkozar for discussion :)

### CODE REVIEW

**Comment by trishume** [comment]
This is neat. Do you think it would be possible someday to have the layering work within the nix expressions? Could you have `baseImage = bashImage` and `bashImage = dockerImage { ... }` so that Nixpkgs could provide a generic baseImage attribute that lives in the nix store that other specialty images could layer on top of?

Also it sounds like with some tuning (figure out why gcc is pulled in) Nix could create the leanest container images of any OS since *exactly* the minimal set of dependencies for any given program would be pulled in. That would be almost micro-kernel like.

**Comment by lethalman** [comment]
@trishume I'm thinking about that and I feel that it **may** be possible. `docker save` would be the derivation output of an image. At this point inside nix, run the docker daemon with custom paths, `docker load`, `docker build` and finally `docker save`. It all depends if the docker daemon may run as a user and with enough privileges for creating his filesystems inside nix.

And yes, docker images with nix, if packages were properly optimized, would deliver the smallest images.

**Comment by datakurre** [comment]
@lethalman That 184 MB for Python sounds strange. I've been getting 36 MB tar.gz and 108.9 MB unpacked "virtual size" already. Since I've still learning Nix as language, I've simply:

```
nix-build python.nix
tar cvz `nix-store -qR result ` result > python.tar.gz
```
Also good to know: If there's no explicit PATH env defined, Docker run seems to add the normal /usr/local/bin:/usr/bin:/bin/etc... so that ``docker run -t -i nixpython python`` should be enough for run.

Then there's that issue of many software expecting /bin/sh or /bin/bash and writable /tmp to exist. And usually one need to add many derivations to PATH. I went far enough to map expression build result into image root with:

```
mkdir tmp
tar cvz --transform="s|^result/||" tmp `nix-store -qR result` result/* > image.tar.gx
```

And with buildEnv that results with /bin et al. having all the expected paths. Probably there are also downsides.

**Comment by lethalman** [comment]
@datakurre that's not only python, but also tools for extending the image later with layers.

**Comment by lethalman** [comment]
@trishume so it seems possible to do with `runInLinuxVM`, however there's a drawback of `docker save/load`. The of the image cannot be chosen when loading, but it can be renamed afterwards. So indeed we can generate layered images. Will do a complete PR.

**Comment by datakurre** [comment]
@lethalman That explains. You figured out quite a way to get data into image without through the filter without ADD or COPY :)

Still, could there be a way to create delta.tar.gz outside Docker (e.g. getting the paths from the base image with docker save) and use a Dockerfile with ``ADD delta.tar.gz /`` instead? According to the docs https://docs.docker.com/engine/reference/builder/#add ADD has bevior of ``tar -x``

    the result is the union of:

    Whatever existed at the destination path and
    The contents of the source tree, with conflicts resolved in favor of “2.” on a file-by-file basis.

**Comment by lethalman** [comment]
@datakurre with a delta it could be possible yes, but it's still another step to do from the outside :(

**Comment by datakurre** [comment]
@lethalman Another option could be to optionally manage image "version" with Nix profiles. Could that make any sense? You would not want that for one one-shot images, but if you'd maintain a long-living production app it might make sense.

**Comment by lethalman** [comment]
@datakurre can you explain better? I didn't understand the image version with Nix profiles...

**Comment by datakurre** [comment]
@lethalman Sorry, I missed your update on deltas. Is there any option, which would work within Docker? (To make it possible to build a Linux Docker image on a Mac with boot2docker.)

About the profiles. I was thinking, if it would make sense to manage named dockered apps using Nix profiles so that it would be possible to create a delta tarball by diffing paths from the previous profile version to the current version. But I didn't think this too much. Probably something, where a bash script works better than a nix expression. And very impure.

**Comment by lethalman** [comment]
@datakurre the problem is `docker build` , so you still need impurity, hence no reason to put other metadata somewhere else... because you need impurity anyway.

I didn't understand you first question...

**Comment by datakurre** [comment]
@lethalman Agreed. I understand this now.

Also the first question was probably a misunderstanding from me. So, would ``nix-docker-merge`` result an image.tar.gz, Dockerfile and the script to create the delta for the base image? So that to build x86 containers, I could call dockerImage at Nix container in VM, copy the results into darwin host and run the script on darwin host?

**Comment by lethalman** [comment]
`nix-docker-merge` creates the Dockerfile, the delta and outputs the saved image. If the VM has x86 nix, yes. Note that `runInLinuxVM` bind mounts the host /nix, also I don't know if it works on darwin.

So if you have an x86 VM, and you create two docker images from there, you can merge the result of course regardless of the architecture. Basically that `nix-docker-merge` has nothing to do with nix, it only uses `nix-hash` :)

**Comment by lethalman** [comment]
Alright :D We have it...

```
with import <nixpkgs> {};

let
  image1 = dockerTools.createImage { drv = bashInteractive; };
  image2 = dockerTools.createImage { drv = python; };
in
  dockerTools.mergeImages { name = "bash-python"; inherit image1 image2; }
```

```
$ nix-build
/nix/store/br5hqbzcqcgv7fy85pp7h91smpia32nb-bash-python.tar.gz
# docker load -i result
# docker history 240c630f46d8
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
240c630f46d8        5 minutes ago       /bin/sh -c #(nop) ADD dir:f576a907f86b772f9b7   68.29 MB            
df1dafed18ab        6 minutes ago                                                       50.16 MB            Imported from -
```

There are two problems:
1. When doing `docker build` I've passed `-t bash-python`, but then `docker load` ignore the name and imported it anonymously.
2. But the real problem is that `docker load` does not print the image id :( hence `mergeImages` will not work with layered images................... that's really something to report to docker.

**Comment by geerds** [comment]
@lethalman I fell in love with you :laughing: 

That's so great! I have to test it

**Comment by domenkozar** [comment]
@geerds don't scare him away :warning:

**Comment by lethalman** [comment]
Just reported the `docker load` problem upstream: https://github.com/docker/docker/issues/18117

**Comment by datakurre** [comment]
:+1:

**Comment by lethalman** [comment]
Alright. For now I've workarounded that issue, and we can only deal with layered images that have a name, because we cannot know easily the image id of the loaded image.

So we can stack a many layers as we want from independent images:
```
with import <nixpkgs> {};

rec {
  image1 = dockerTools.createImage { drv = bash; };
  image2 = dockerTools.createImage { drv = python; };
  image3 = dockerTools.createImage { drv = perl; };
  bash-python = dockerTools.mergeImages { name = "bash-python"; baseImage = image1; image = image2; };
  bash-python-perl = dockerTools.mergeImages { name = "bash-python-perl"; baseImage = bash-python; image = image3; };
}
```

Do `nix-build -A bash-python-perl` et voila!

Going to add also some other useful tools, like converting a flattened image to an image with one layer, and viceversa.

**Comment by lethalman** [comment]
@edolstra do you feel we can merge this? It's just tools, no harm for the build farm.

**Comment by lethalman** [comment]
Docker is taking up so much disk space (like over 2gb...) that I'm thinking of implementing the packing format of the images by myself. It's quite simple, and there will be no need of `runInLinuxVM`, completely pure.

https://github.com/docker/docker/blob/master/image/spec/v1.md

And they even reimplemented nix-hash: https://github.com/docker/docker/blob/master/pkg/tarsum/tarsum_spec.md

**Comment by lethalman** [comment]
There we go, cleaned up of all the docker impure and huge evilness. It's all pure nix, we generate compatible docker images.

```
with import ./. {};

rec {
  image1 = dockerTools.mkImage { drv = bash; };
  layeredImage1 = dockerTools.toLayeredImage { name = "bash"; image = image1; };
  image2 = dockerTools.mkImage { drv = python; };
  image3 = dockerTools.mkImage { drv = perl; };
  bash-python = dockerTools.mergeImages { name = "bash-python"; baseImage = layeredImage1; image = image2; };
  bash-python-perl = dockerTools.mergeImages { name = "bash-python-perl"; baseImage = bash-python; image = image3; };
}
```

```
# docker load -i result
# docker history bash-python-perl
IMAGE               CREATED               CREATED BY          SIZE
2fbfd365736b        45.916921 years ago                       145.2 MB
c23bc184169f        45.916921 years ago                       68.29 MB
8ed4ee623119        45.916921 years ago                       40.8 MB
# docker run -t -i bash-python-perl /nix/store/rfcsnszaw2jlkqdgg65h3phss8xa0mlp-python-2.7.10/bin/python
Python 2.7.10 (default, Jan 01 1970, 00:00:01) 
[GCC 4.9.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>>

**Comment by lethalman** [comment]
New syntax:

```
with import <nixpkgs> {};

rec {
  image1 = dockerTools.mkTarball { drv = bash; };
  layeredImage1 = dockerTools.build { name = "bash"; addTarball = image1; };
  image2 = dockerTools.mkTarball { drv = python; };
  image3 = dockerTools.mkTarball { drv = perl; };
  bash-python = dockerTools.build { name = "bash-python"; fromImage = layeredImage1; addTarball = image2; };
  bash-python-perl = dockerTools.build { name = "bash-python-perl"; fromImage = bash-python; addTarball = image3; };

  just-python = dockerTools.build { name = "just-python"; config = { Cmd = [ "${python}/bin/${python.executable}" ]; }; };
}
```

But now we also support arbitrary config and its dependencies! Look `just-python`:

```
$ nix-build -A just-python
# docker load -i result
# docker run -t -i just-python
Python 2.7.10 (default, Jan 01 1970, 00:00:01) 
[GCC 4.9.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

We basically have a partial `docker build` working. Other stuff like handling multiple images ecc. are not working of course.

**Comment by lethalman** [comment]
I'm going to merge this in a few days, unless somebody is contrary.

**Comment by domenkozar** [comment]
@lethalman would be really awesome to have this in nixpkgs docs

**Comment by lethalman** [comment]
@domenkozar of course, I still have some ideas to make this interface even more pleasant... once finished I will write docs.

**Comment by domenkozar** [comment]
:cake:

**Comment by trishume** [comment]
Holy crap, composable docker builds reimplemented with Nix in <100 lines of Nix expressions. They look really easy to use. This is really cool work @lethalman!

I wonder how well hydra could do as a container build infrastructure. I worked at a company that was scaling up a containerized deployment process and their container build system was built from scratch. It later became a bottleneck since it wasn't distributed. Using Hydra's distributed builds could have solved that problem and they wouldn't have even had to write the build infrastructure that they did.

Might even be able to plug a deployment framework up to Hydra to ship the built containers to a cluster.

**Comment by lethalman** [comment]
@trishume and the advantages:
1. No need to worry about the "context"
2. No need to chain `RUN && && && &&` to the death, nix only does one layer per build.
3. It takes care of making an actual diff with the base image.
4. As usual nix, you don't need to sort out dependencies manually.
5. No need to mess with your system docker.
6. Smaller, even smaller when packages will have proper multi outputs.

**Comment by lethalman** [comment]
There is a problem with this approach however... it's not possible change special permissions to files or change ownership.

I think for that case `runInDockerVM` must be used :(

**Comment by trishume** [comment]
Can this approach control exposed ports and control environment variables like a Dockerfile? If the answer is "not yet but possible" that can definitely be a future addition, I don't want to make it sound like I don't think this should be merged yet.

**Comment by offlinehacker** [comment]
This is defintelly interesting and +1 for merging this, but in our deployments we are still planning to take another approach with distributed and shared nix store, where you don't even need any image, docker registry and other components, you just run software you have in nix store and that's it.

**Comment by offlinehacker** [comment]
@trishume exposed ports can be always controller with `docker run -p ...`

**Comment by lethalman** [comment]
@trishume this is already possible, see the `config` variable.

@offlinehacker of course, as long as you don't have to deal with foreign cloud environments that's the perfect solution ;)

**Comment by datakurre** [comment]
This will be a nice solution for those (me included), who have already docker deployments set up, but would like to try transition to nix based builds. Once all builds are nix'd, moving away from docker is easier.

**Comment by lethalman** [comment]
Ok so this is a new example:

```
buildImage { name = "python-link";
                     run = ''
                       mkdir bin
                       ln -s ${python}/bin/${python.executable} bin/python
                     '';
                     config = { Cmd = [ "/bin/python" ]; }; };
```

The run script will be converted as derivation, its deps will be included (like for the config) and the contents of the run script will be inside the image. Just convenience.

So it should be complete. Accepting reviews before merging, like missing use cases ecc..

**Comment by lethalman** [comment]
Alright... I've finally managed to get something from the docker library hub written in nix. I started from something simple, redis.

So this is the dockerfile: https://github.com/docker-library/redis/blob/8929846148513a1e35e4212003965758112f8b55/3.0/Dockerfile
And this is our nix expression: https://github.com/NixOS/nixpkgs/blob/f5412cfc12d93c4bf9a8750bba90e4fcd7f3552f/pkgs/top-level/docker-library.nix#L29

Also it's possible to `redis.override { fromImage = utils; };`, which makes it easy to create containers for production or for development.

That opens up to `pkgs/top-level/docker-library.nix`, hence `pkgs.dockerLibrary`. However we probably won't enable those for hydra.

Note the `runAsRoot` runs in a linux vm, so it requires kvm, and uses overlayfs, which is the only simple "union" fs available by default in our kernel. It's still not complete as the instrumentation doesn't take in account deleted files, but that's easy to do.

**TODO**:

[ ] - Short term: use `tarsum` instead of our `nix-hash`. Our `nix-hash` works on directory but we need to sum the layer.tar.
[ ] - Short term: flatten an image to a single tarball, i.e. `docker export` -> `flattenImage`.
[ ] - Long term: add instrumentation to manipulate tarballs. Most of docker impure commands are useradd, creating root files, ecc., which with some tooling they may be done by tweaking directly the layer.tar instead of the filesystem, thus avoiding even the kvm, making it fully pure.

That opens up to `pkgs/top-level/docker-library.nix`, hence `pkgs.dockerLibrary`. However we probably won't enable those for hydra.

**Comment by lethalman** [comment]
After some thinking, I believe it's better to avoid adding a docker library in our nixpkgs. If people is interested, it can be done in a separate repository. Time to write docs I guess...

**Comment by puffnfresh** [comment]
@lethalman I'm very interested, I've wanted exactly this for a long time and you've done an amazing job :fireworks: 

I think it would be great in nixpkgs but will follow along wherever it is.

**Comment by lethalman** [comment]
So I added a `fetchFromRegistry`, too bad docker hub returns non-deterministic json metadata: https://github.com/docker/hub-feedback/issues/488

Hope they will tell me that only happens with empty tars so we may add some workaround...... :(

**Comment by lethalman** [comment]
Done documenting the API. There are some hidden features but are not important to document for now.

**Comment by trishume** [comment]
The docs are awesome!

A couple questions I have that I couldn't answer from the docs:
- If I use Nix to build a 3GB image and then 3 independent 10mb layers on top of that image, will it use more like 3GB + 3x10mb or 4x3GB + 3x10mb. I.E is the base image copied in the nix store when you add a layer to it?
- Is there any reason I would want to layer on top of a Docker Pull or is it more for whole containers. If I take a debian image and then add Python from Nix as a layer, won't none of the Debian stuff get used since the Nix Python will include and use its dependency closure? I definitely see the utility for things like pre-built redis containers though.

These questions may well be answered, but I might not have read hard enough.

**Comment by lethalman** [comment]
It will use space in the store once for the base image, then temp space for unpacking, the new layer in the store and the result in the store. All this is probably still less than what's required by docker. All can be garbage collected after you get the result.

For example you want a debian as base and nix software, because you may want debian tools and environment or some apt packages. Or some base fancy openstack image, wordpress, rails, anything that is not in nixpkgs or hard to package, then layer your app on top of it. Because your sysadm knows debian and provides base debian images... Whatever.

**Comment by trishume** [comment]
Oh I see, so you could use a base container like a fancy apache java servlet whatever with monitoring and log forwarding and all sorts of stuff. Then you can build your java app with nix and layer that on top, then just use a shell script to connect your Java app built with Nix up to whatever servlet thingy comes with the container.

The key being that the Nix layer won't interact with anything in the base layer on its own, but base layer services can still run, and you can use a build script to connect up your nix build to things present in the base image.

It might be helpful to have an example use case in the docs. For the purpose of marketing how cool and useful this stuff is.

**Comment by anderspapitto** [comment]
I would also like to see this land in the main nixpkgs repo.

**Comment by anderspapitto** [comment]
also, is it still possible to create multiple layers? It at least seems like mergeLayers doesn't exist any more, and your last syntax example is only a single layer.

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+        
+        mkdir temp
+        cp ${layer}/* temp/
+        chmod +w temp/*
+        
+        echo Adding dependencies
+        tar -tf temp/layer.tar >> baseFiles
+        tar -tf "$mergedTarball" | grep -v ${layer} > layerFiles
+        sed -i -e 's|^[\./]\+||' baseFiles layerFiles
+        comm <(sort -n baseFiles|uniq) <(sort -n layerFiles|uniq) -1 -3 > newFiles
```
Comment (anderspapitto) [inline pkgs/build-support/docker/default.nix]: I was getting some errors from comm when I tried this out, saying that file 1 was not sorted. Removing the -n flag to sort made that error disappear, so unless it's there for a particular reason you may want to remove it

Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Perhaps you are using darwin or something like that? I can remove it yes, the specific kind of order is not important, as long as they have some order. Thanks for the feedback.

**Comment by lethalman** [comment]
@anderspapitto every `buildImage` adds one layer. Hence 5 `buildImage`, 5 layers. While docker adds a layer for each `Dockerfile` command.

**Comment by globin** [comment]
Very nice work :cake: 

I'm also :+1: to add this to nixpkgs directly.

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (anderspapitto) [inline pkgs/build-support/docker/default.nix]: I'm on straight-up NixOS. Here's a small repro of what I see

```
[anders@gurney:~/playground/nix-docker]
$ nix-build gcc.nix -A image
these derivations will be built:
  /nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv
building path(s) ‘/nix/store/cls0x1v8g6d5zivp00cqxiddiyx9nsi9-gcc.tar.gz’
Adding dependencies
comm: file 1 is not in sorted order
builder for ‘/nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv’ failed with exit code 1
error: build of ‘/nix/store/hdchj8q53khs4lz88c219hflgghnzxil-gcc.tar.gz.drv’ failed
[anders@gurney:~/playground/nix-docker]
$ cat gcc.nix
with import <nixpkgs> { };
with dockerTools;

rec {
  image = buildImage {
    name = "gcc";
    contents = [ gcc distcc ];
    config = {
      Cmd = [ "${gcc}/bin/gcc" ];
    };
  };
}
```

I'm on nixpkgs 07ab8c6c01a2ca74baf5fde2fad45833ccf09bba, with this change cherry-picked on top

**Comment by anderspapitto** [comment]
one extra thought - would it be possible to avoid the expensive filesystem diffing in some cases, for example if the final layer only changed the 'config' attribute, and didn't introduce any new deps? currently it's a little painful to play around with, since changing an exposed port or a command line flag is a minute+ operation

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: `contents` must be a single derivation not a list. So you may want `buildEnv { name="foo"; paths = [ gcc distcc ]; }`.
The weird thing is that it built for me :O

**Comment by lethalman** [comment]
@anderspapitto note that `config` may as well introduce dependencies (forgot to mention in the docs). However if the dependencies tarball is empty, yes we can avoid the diff only in that case.

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Another question, are you building with chroot enabled?

Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Another question, are you building with chroot enabled?

Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Btw I will just add `--nocheck-order`. Would it work for you with this flag?

Comment (lethalman) [inline pkgs/build-support/docker/default.nix]: Will also add the possibility to specify a list of contents.

**Comment by couchemar** [comment]
Is there  proper way to add file to image? I want to create container with Nginx with my nginx.conf:

```
with import /home/andrey/development/nixpkgs {};
with dockerTools;

let
    nginxConf = writeText "nginx.conf" (builtins.readFile ./nginx.conf);
in

buildImage {
  name = "nginx-container";
  contents = pkgs.nginx;

  config = {
    Cmd = [ "nginx -c ${nginxConf}" ];
    ExposedPorts = {
      "80/tcp" = {};
    };
  };
}
```
Image is successfully created, but when I run it, I got an error:
```
docker run -t -i nginx-container
exec: "nginx -c /nix/store/31hslif7vdy2hmc43wagc3nnhh6lzicj-nginx.conf": stat nginx -c /nix/store/31hslif7vdy2hmc43wagc3nnhh6lzicj-nginx.conf: no such file or directory
```

**Comment by domenkozar** [comment]
try `contents =  [pkgs.nginx nginxConf ];`

**Comment by lethalman** [comment]
@couchemar that seems like a bug... will investigate thanks.

**Comment by lethalman** [comment]
@couchemar ah no it's correct, `Cmd` is an array of arguments... split it

**Comment by couchemar** [comment]
@lethalman Thank you! This works now.

**Comment by couchemar** [comment]
It seems that `runAsRoot` does not knew about `export`:
```
with import ../. {};
with dockerTools;

buildImage {
    name = "shadow-basic";

    runAsRoot = ''
      ${shadowSetup}
    '';
}
```
 ```
$ nix-build
these derivations will be built:
  /nix/store/ka6jpv3n41lxrp1lbc7qaai12211z3xs-docker-layer.drv
  /nix/store/d670hzfpy5kq9pmw0i60s1y2m0flxfgi-runtime-deps.drv
  /nix/store/l4ndcghxf6g76lxyyd6awyay9mppw3mx-shadow-basic-deps.tar.gz.drv
  /nix/store/khnrm6gw570lvcp19c01xlbxp0lshl86-shadow-basic.tar.gz.drv
building path(s) ‘/nix/store/6nmabpmblgj0r0hsd1ii9bxff2pk2pgn-docker-layer’
Formatting '/nix/store/6nmabpmblgj0r0hsd1ii9bxff2pk2pgn-docker-layer/disk-image.qcow2', fmt=qcow2 size=1073741824 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
loading kernel modules...
mounting Nix store...
mounting host's temporary directory...
starting stage 2 (/nix/store/yfl34z6w8s3c54b9994564alsfggqxwp-vm-run-stage2)
mke2fs 1.42.13 (17-May-2015)
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: 38e0b905-6667-43dd-a9cb-dc63d1dd1e27
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Writing superblocks and filesystem accounting information: done

chroot: failed to run command 'export': No such file or directory
[    6.434359] reboot: Power down
builder for ‘/nix/store/ka6jpv3n41lxrp1lbc7qaai12211z3xs-docker-layer.drv’ failed with exit code 127
cannot build derivation ‘/nix/store/khnrm6gw570lvcp19c01xlbxp0lshl86-shadow-basic.tar.gz.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/khnrm6gw570lvcp19c01xlbxp0lshl86-shadow-basic.tar.gz.drv’ failed
```

**Comment by lethalman** [comment]
@couchemar put `#!/bin/sh` on top...

**Comment by couchemar** [comment]
@lethalman still error...
```
$ cat default.nix 
with import ../. {};
with dockerTools;

buildImage {
    name = "shadow-basic";

    runAsRoot = ''
      #!/bin/sh
      ${shadowSetup}
    '';
}
```
```
$ nix-build
these derivations will be built:
  /nix/store/2s97hmi0x61mdl4mv1kh1kckxv4p5rqx-docker-layer.drv
  /nix/store/mfakz29ymvdhcklpzdsb6rqx3c68wvj4-runtime-deps.drv
  /nix/store/s14lds43pa7gvvm32pbbz03x1bc0nzxq-shadow-basic-deps.tar.gz.drv
  /nix/store/jx7bs22w4fnamm71nb7bcw27g9g3ampm-shadow-basic.tar.gz.drv
building path(s) ‘/nix/store/2kkfz3nczbsrqayp11l58k1jsjy47jk1-docker-layer’
Formatting '/nix/store/2kkfz3nczbsrqayp11l58k1jsjy47jk1-docker-layer/disk-image.qcow2', fmt=qcow2 size=1073741824 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
loading kernel modules...
mounting Nix store...
mounting host's temporary directory...
starting stage 2 (/nix/store/yfl34z6w8s3c54b9994564alsfggqxwp-vm-run-stage2)
mke2fs 1.42.13 (17-May-2015)
Creating filesystem with 262144 4k blocks and 65536 inodes
Filesystem UUID: ccb2d9d4-9711-44a5-af94-be074370afff
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Writing inode tables: done                            
Writing superblocks and filesystem accounting information: done

bash: cannot set terminal process group (-1): Inappropriate ioctl for device
bash: no job control in this shell
bash-4.3# exit
/nix/store/h84kc62s87mknjrln5rd6s332zfxfrk5-stdenv/setup: line 867: /etc/pam.d/other: No such file or directory
[    6.408975] reboot: Power down
builder for ‘/nix/store/2s97hmi0x61mdl4mv1kh1kckxv4p5rqx-docker-layer.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/jx7bs22w4fnamm71nb7bcw27g9g3ampm-shadow-basic.tar.gz.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/jx7bs22w4fnamm71nb7bcw27g9g3ampm-shadow-basic.tar.gz.drv’ failed

```

**Comment by lethalman** [comment]
ah ok that's different @couchemar will fix it

**Comment by lethalman** [comment]
fixed and rebased on current master

**Comment by lethalman** [comment]
@rvl ok the example was misleading because `runAsRoot` required directly an executable. Now it can be a string and your code should work.

**Comment by lethalman** [comment]
@anderspapitto optimized the diffing part, but note the most expensive operations are due to unpacking and repacking the whole image.

**Comment by rvl** [comment]
Thanks @lethalman, [it works well](https://gist.github.com/rvl/c1ca657882a883195f49#file-test-docker-nix-L67). Is there anything we can do to help getting this merged?

I reviewed the code (or tried to). The only questions I can come up with are:
 - how long will the docker hub registry `/v1` continue to work?
 - is there any way of calculating a timestamp for a derivation which isn't 1970?
 - is the `result` let-binding needed in `buildImage`? (seems to work without it)
 - is an end-to-end test possible? ... e.g. build an image, load into docker, check that it runs.

**Comment by lethalman** [comment]
@rvl thanks for testing.

- Well, they don't have a v2 so... :)
- Derivations have to be reproducible. You have to think they are addressed by hash, not by time. That's for any derivation in nix, not only docker images. I understand that's ugly, it's as ugly as running uname on NixOS.
- It's handy just in case we need to tweak the result further. If in the future you have to re-indent, that would break the blame.
- Yes, it's possible with runInLinuxVM but also like with any other nixos test. There's even a docker test already for nixos.

**Comment by lethalman** [comment]
As for the last bits todo:

- Allow contents to be a list.
- Currently in some code paths, when copying contents, there's a chance it fails due read-only files of different nixbld users. It should suffice to do a chown -R a+w after copying stuff.

**Comment by lethalman** [comment]
I should have addressed the last bits too, and the docs about the default values for `pullImage`.

I've rebased on top of current master. Asking for some other testing now, then I'll merge.

**Comment by rvl** [comment]
Great, it still works for me. A list of derivations in `contents` also works for me. Can I suggest rewording the sentence in `doc/functions.xml:405` to "contents is one or more derivations that will be copied in the new layer of the resulting image."

**Comment by lethalman** [comment]
@rvl thanks for testing. How should it be reworded?

**Comment by rvl** [comment]
Instead of "contents is a derivation that will be copied in the new layer of the resulting image", try "contents is one or more derivations that will be copied in the new layer of the resulting image." I think that would suggest using a list if necessary.

**Comment by jgillich** [comment]
What's the status of this?

**Comment by lethalman** [comment]
@jgillich mergeable. I'm going to rebase, retest and merge right now.

**Comment by jgillich** [comment]
Awesome! :dancers:

**Comment by lethalman** [comment]
Tested `pullImage`, sha is almost always wrong. I think docker changed again something :(

**Comment by lethalman** [comment]
Nvm it's now somewhat consistent. I guess there's still some non-determinism somewhere in `pullImage`, but we'll see with time.

**Comment by lethalman** [comment]
Merged. No reason to keep this unmerged, it can be improved over time and already does a very good job I think.

**Comment by datakurre** [comment]
:1:

**Comment by cstrahan** [comment]
Wow, this is pretty awesome looking!

**Comment by datakurre** [comment]
@lethalman Would you know, is something special required for hydra on nixos to be able to build image with runAsRoot (requiring kvm)? Images without buildAsRoot build fine. Images with runAsRoot build fine as a user, but not by hydra (on the same machine).

**Comment by datakurre** [comment]
@lethalman Sorry for bothering. It seems that ``supportedFeatures = [ "kvm" ];`` for nix.buildMachine did it.
